src/mlpack/methods/ann/rbm/spike_slab_rbm_impl.hpp

```diff
@@ -0,0 +1,311 @@
+/**
+ * @file spike_slab_rbm_impl.hpp
+ * @author Kris Singh
+ * @author Shikhar Jaiswal
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license. You should have received a copy of the
+ * 3-clause BSD license along with mlpack. If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_RBM_SPIKE_SLAB_RBM_IMPL_HPP
+#define MLPACK_METHODS_ANN_RBM_SPIKE_SLAB_RBM_IMPL_HPP
+
+#include "rbm.hpp"
+
+#include <mlpack/methods/ann/activation_functions/logistic_function.hpp>
+#include <mlpack/methods/ann/activation_functions/softplus_function.hpp>
+
+
+namespace mlpack {
+namespace ann {
+
+template<
+  typename InitializationRuleType,
+  typename DataType,
+  typename PolicyType
+>
+template<typename Policy>
+typename std::enable_if<std::is_same<Policy, SpikeSlabRBM>::value, void>::type
+RBM<InitializationRuleType, DataType, PolicyType>::Reset()
+{
+  size_t shape = (visibleSize * hiddenSize * poolSize) + visibleSize + hiddenSize;
+  parameter.set_size(shape, 1);
+  positiveGradient.set_size(shape, 1);
+  negativeGradient.set_size(shape, 1);
+  tempNegativeGradient.set_size(shape, 1);
+  negativeSamples.set_size(visibleSize, batchSize);
+  visibleMean.set_size(visibleSize, 1);
+  spikeMean.set_size(hiddenSize, 1);
+  spikeSamples.set_size(hiddenSize, 1);
+  slabMean.set_size(poolSize, hiddenSize);
+
+  // Weight shape D * K * N
+  weightCube = arma::Cube<ElemType>(parameter.memptr(), visibleSize, poolSize,
+      hiddenSize, false, false);
+  // spike bias shape N * 1
+  spikeBias = DataType(parameter.memptr() + weight.n_elem, hiddenSize, 1,
+      false, false);
+  // visible penalty 1 * 1 => D * D(when used)
+  visiblePenalty = DataType(parameter.memptr() + weight.n_elem +
+      spikeBias.n_elem, 1, 1, false, false);
+
+  parameter.zeros();
+  positiveGradient.zeros();
+  negativeGradient.zeros();
+  tempNegativeGradient.zeros();
+  initializeRule.Initialize(parameter, parameter.n_elem, 1);
+
+  reset = true;
+}
+
+template<
+  typename InitializationRuleType,
+  typename DataType,
+  typename PolicyType
+>
+template<typename Policy>
+typename std::enable_if<std::is_same<Policy, SpikeSlabRBM>::value, double>::type
+RBM<InitializationRuleType, DataType, PolicyType>::FreeEnergy(
+    arma::Mat<ElemType>&& input)
+{
+  ElemType freeEnergy = 0.5 * visiblePenalty(0) * arma::accu(input.t() * input);
```

Comment:
We could also use ``arma::dot(input, input)`` instead of `` arma::accu(input.t() * input)`` here, it's not really shorter.